src/Access/ReplicatedAccessStorage.cpp

```diff
@@ -437,185 +504,129 @@ void ReplicatedAccessStorage::refreshEntities(const zkutil::ZooKeeperPtr & zooke
-    for (const auto & pair : entries_by_id)
+    if (all)
     {
-        const UUID & entity_uuid = pair.first;
-        if (!entity_uuids.contains(entity_uuid))
-            entities_to_remove.push_back(entity_uuid);
+        /// all=true means we read & parse all access entities from ZooKeeper.
+        std::vector<std::pair<UUID, AccessEntityPtr>> entities;
+        for (const auto & uuid : entity_uuids)
+        {
+            if (auto entity = tryReadEntityFromZooKeeper(zookeeper, uuid))
+                entities.emplace_back(uuid, entity);
+        }
+        memory_storage.setAll(entities);
     }
-    for (const auto & entity_uuid : entities_to_remove)
-        removeEntityNoLock(entity_uuid);
-
-    /// Locally add entities that were added to ZooKeeper
-    for (const auto & entity_uuid : entity_uuids)
+    else
     {
-        const auto it = entries_by_id.find(entity_uuid);
-        if (it == entries_by_id.end())
-            refreshEntityNoLock(zookeeper, entity_uuid);
+        /// all=false means we read & parse only new access entities from ZooKeeper.
+        memory_storage.removeAllExcept(entity_uuids);
+        for (const auto & uuid : entity_uuids)
+        {
+            if (!memory_storage.exists(uuid))
+                refreshEntityNoLock(zookeeper, uuid);
+        }
     }
 
     LOG_DEBUG(getLogger(), "Refreshing entities list finished");
 }
 
+
 void ReplicatedAccessStorage::refreshEntity(const zkutil::ZooKeeperPtr & zookeeper, const UUID & id)
 {
+    LOG_DEBUG(getLogger(), "Refreshing entity {}", toString(id));
+
+    auto entity = tryReadEntityFromZooKeeper(zookeeper, id);
+
     std::lock_guard lock{mutex};
-    refreshEntityNoLock(zookeeper, id);
+
+    if (entity)
+        setEntityNoLock(id, entity);
+    else
+        removeEntityNoLock(id);
 }
 
 void ReplicatedAccessStorage::refreshEntityNoLock(const zkutil::ZooKeeperPtr & zookeeper, const UUID & id)
 {
     LOG_DEBUG(getLogger(), "Refreshing entity {}", toString(id));
 
+    auto entity = tryReadEntityFromZooKeeper(zookeeper, id);
+    if (entity)
+        setEntityNoLock(id, entity);
+    else
+        removeEntityNoLock(id);
+}
+
+AccessEntityPtr ReplicatedAccessStorage::tryReadEntityFromZooKeeper(const zkutil::ZooKeeperPtr & zookeeper, const UUID & id) const
+{
     const auto watch_entity = [watched_queue = watched_queue, id](const Coordination::WatchResponse & response)
     {
         if (response.type == Coordination::Event::CHANGED)
             [[maybe_unused]] bool push_result = watched_queue->push(id);
     };
+
     Coordination::Stat entity_stat;
     const String entity_path = zookeeper_path + "/uuid/" + toString(id);
     String entity_definition;
-    const bool exists = zookeeper->tryGetWatch(entity_path, entity_definition, &entity_stat, watch_entity);
-    if (exists)
+    bool exists = zookeeper->tryGetWatch(entity_path, entity_definition, &entity_stat, watch_entity);
+    if (!exists)
```

Comment:
```suggestion
    if (exists = zookeeper->tryGetWatch(entity_path, entity_definition, &entity_stat, watch_entity);
        !exists)
```

Reply:
I think this is a bit less readable because the scope of the variable `exists` doesn't really matter here.